pvlib/modelchain.py

```diff
@@ -1005,77 +1160,88 @@ def complete_irradiance(self, weather):
         >>> mc = ModelChain(my_system, my_location)  # doctest: +SKIP
         >>> mc.run_model(my_weather)  # doctest: +SKIP
         """
+        self._check_multiple_input(weather)
         self.weather = weather
+        self._assign_times()
+        self.results.solar_position = self.location.get_solarposition(
+            self.times, method=self.solar_position_method)
+
+        if isinstance(weather, tuple):
+            for w in self.weather:
+                self._complete_irradiance(w)
+        else:
+            self._complete_irradiance(self.weather)
 
-        self.solar_position = self.location.get_solarposition(
-            self.weather.index, method=self.solar_position_method)
+        return self
 
-        icolumns = set(self.weather.columns)
+    def _complete_irradiance(self, weather):
+        icolumns = set(weather.columns)
         wrn_txt = ("This function is not safe at the moment.\n" +
```

Comment:
warning only needs to be emitted once, so should go before the `if/else` in `complete_irradiance`.

Reply:
It would be good to only emit once, but because the warning is only emitted for certain combinations of columns (only those where `'dni'` is present) it is possible that you could get no warning, but we don't know until we look at each frame. Because of that it cannot go before the `if/else` in `complete_irradiance()`.

I could add a return value to the helper function indicating whether a warning should be emitted (name of column added, or a boolean flag), and some logic in `complete_irradiance()` that determines whether to warn.

Reply:
good point. ignore my comment.